File: packages/python/glin_sdk/contracts/escrow.py
"""GenericEscrow contract wrapper"""

from typing import Optional, List
from substrateinterface import SubstrateInterface, Keypair, ContractInstance, ContractCode
from .types import (
    Agreement,
    Milestone,
    MilestoneStatus,
    CreateAgreementParams,
    ContractResult
)


class EscrowContract:
    """
    Wrapper for GenericEscrow smart contract interactions

    Provides methods to interact with the GenericEscrow contract
    for milestone-based payment escrow with dispute resolution.
    """

    def __init__(
        self,
        substrate: SubstrateInterface,
        contract_address: Optional[str] = None,
        keypair: Optional[Keypair] = None
    ):
        """
        Initialize escrow contract wrapper

        Args:
            substrate: Connected SubstrateInterface instance
            contract_address: GenericEscrow contract address
            keypair: Keypair for signing transactions
        """
        self.substrate = substrate
        self.contract_address = contract_address
        self.keypair = keypair
        self.contract: Optional[ContractInstance] = None

        if contract_address:
            self._load_contract(contract_address)

    def _load_contract(self, address: str) -> None:
        """Load contract instance with metadata"""
        # In production, load metadata from file or URL
        # For now, we'll handle contract calls manually
        self.contract_address = address

    def set_contract_address(self, address: str) -> None:
        """Set or update contract address"""
        self.contract_address = address
        self._load_contract(address)

    def set_keypair(self, keypair: Keypair) -> None:
        """Update the keypair for transactions"""
        self.keypair = keypair

    async def create_agreement(
        self,
        params: CreateAgreementParams,
        gas_limit: Optional[int] = None
    ) -> ContractResult:
        """
        Create a new escrow agreement

        Args:
            params: Agreement parameters
            gas_limit: Optional gas limit override

        Returns:
            ContractResult with agreement_id on success

        Example:
            >>> params = CreateAgreementParams(
            ...     provider="5Provider...",
            ...     milestone_descriptions=["Design", "Development"],
            ...     milestone_amounts=["500000000000000000000", "1500000000000000000000"],
            ...     milestone_deadlines=[1234567890, 1234567890],
            ...     dispute_timeout=1234567890,
            ...     value="2000000000000000000000"
            ... )
            >>> result = await escrow.create_agreement(params)
        """
        if not self.contract_address:
            return ContractResult(success=False, error="Contract not initialized")
        if not self.keypair:
            return ContractResult(success=False, error="Keypair not set")

        try:
            # Build contract call
            call = self.substrate.compose_call(
                call_module='Contracts',
                call_function='call',
                call_params={
                    'dest': self.contract_address,
                    'value': int(params.value),
                    'gas_limit': {'ref_time': gas_limit or 100_000_000_000, 'proof_size': 100_000},
                    'storage_deposit_limit': None,
                    'data': self._encode_call(
                        'create_agreement',
                        params.provider,
                        params.milestone_descriptions,
                        params.milestone_amounts,
                        params.milestone_deadlines,
                        params.dispute_timeout,
                        params.oracle
                    )
                }
            )

            # Create and submit extrinsic
            extrinsic = self.substrate.create_signed_extrinsic(
                call=call,
                keypair=self.keypair
            )

            receipt = self.substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)

            if receipt.is_success:
                # Extract agreement_id from events
                agreement_id = self._extract_agreement_id(receipt)
                return ContractResult(success=True, data=agreement_id)
            else:
                return ContractResult(success=False, error="Transaction failed")

        except Exception as e:
            return ContractResult(success=False, error=str(e))

    async def complete_milestone(
        self,
        agreement_id: str,
        milestone_index: int,
        gas_limit: Optional[int] = None
    ) -> ContractResult:
        """
        Mark a milestone as completed (by provider)

        Args:
            agreement_id: Agreement ID
            milestone_index: Index of the milestone
            gas_limit: Optional gas limit override

        Returns:
            ContractResult
        """
        if not self.contract_address:
            return ContractResult(success=False, error="Contract not initialized")
        if not self.keypair:
            return ContractResult(success=False, error="Keypair not set")

        try:
            call = self.substrate.compose_call(
                call_module='Contracts',
                call_function='call',
                call_params={
                    'dest': self.contract_address,
                    'value': 0,
                    'gas_limit': {'ref_time': gas_limit or 50_000_000_000, 'proof_size': 50_000},
                    'storage_deposit_limit': None,
                    'data': self._encode_call('complete_milestone', agreement_id, milestone_index)
                }
            )

            extrinsic = self.substrate.create_signed_extrinsic(call=call, keypair=self.keypair)
            receipt = self.substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)

            return ContractResult(success=receipt.is_success, error=None if receipt.is_success else "Transaction failed")

        except Exception as e:
            return ContractResult(success=False, error=str(e))

    async def approve_and_release(
        self,
        agreement_id: str,
        milestone_index: int,
        gas_limit: Optional[int] = None
    ) -> ContractResult:
        """
        Approve milestone and release funds (by client or oracle)

        Args:
            agreement_id: Agreement ID
            milestone_index: Index of the milestone
            gas_limit: Optional gas limit override

        Returns:
            ContractResult
        """
        if not self.contract_address:
            return ContractResult(success=False, error="Contract not initialized")
        if not self.keypair:
            return ContractResult(success=False, error="Keypair not set")

        try:
            call = self.substrate.compose_call(
                call_module='Contracts',
                call_function='call',
                call_params={
                    'dest': self.contract_address,
                    'value': 0,
                    'gas_limit': {'ref_time': gas_limit or 50_000_000_000, 'proof_size': 50_000},
                    'storage_deposit_limit': None,
                    'data': self._encode_call('approve_and_release', agreement_id, milestone_index)
                }
            )

            extrinsic = self.substrate.create_signed_extrinsic(call=call, keypair=self.keypair)
            receipt = self.substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)

            return ContractResult(success=receipt.is_success, error=None if receipt.is_success else "Transaction failed")

        except Exception as e:
            return ContractResult(success=False, error=str(e))

    async def raise_dispute(
        self,
        agreement_id: str,
        milestone_index: int,
        gas_limit: Optional[int] = None
    ) -> ContractResult:
        """
        Raise a dispute for a milestone

        Args:
            agreement_id: Agreement ID
            milestone_index: Index of the milestone
            gas_limit: Optional gas limit override

        Returns:
            ContractResult
        """
        if not self.contract_address:
            return ContractResult(success=False, error="Contract not initialized")
        if not self.keypair:
            return ContractResult(success=False, error="Keypair not set")

        try:
            call = self.substrate.compose_call(
                call_module='Contracts',
                call_function='call',
                call_params={
                    'dest': self.contract_address,
                    'value': 0,
                    'gas_limit': {'ref_time': gas_limit or 50_000_000_000, 'proof_size': 50_000},
                    'storage_deposit_limit': None,
                    'data': self._encode_call('raise_dispute', agreement_id, milestone_index)
                }
            )

            extrinsic = self.substrate.create_signed_extrinsic(call=call, keypair=self.keypair)
            receipt = self.substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)

            return ContractResult(success=receipt.is_success, error=None if receipt.is_success else "Transaction failed")

        except Exception as e:
            return ContractResult(success=False, error=str(e))

    async def resolve_dispute(
        self,
        agreement_id: str,
        milestone_index: int,
        release_to_provider: bool,
        gas_limit: Optional[int] = None
    ) -> ContractResult:
        """
        Resolve a dispute (by oracle or after timeout)

        Args:
            agreement_id: Agreement ID
            milestone_index: Index of the milestone
            release_to_provider: Whether to release funds to provider
            gas_limit: Optional gas limit override

        Returns:
            ContractResult
        """
        if not self.contract_address:
            return ContractResult(success=False, error="Contract not initialized")
        if not self.keypair:
            return ContractResult(success=False, error="Keypair not set")

        try:
            call = self.substrate.compose_call(
                call_module='Contracts',
                call_function='call',
                call_params={
                    'dest': self.contract_address,
                    'value': 0,
                    'gas_limit': {'ref_time': gas_limit or 50_000_000_000, 'proof_size': 50_000},
                    'storage_deposit_limit': None,
                    'data': self._encode_call('resolve_dispute', agreement_id, milestone_index, release_to_provider)
                }
            )

            extrinsic = self.substrate.create_signed_extrinsic(call=call, keypair=self.keypair)
            receipt = self.substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)

            return ContractResult(success=receipt.is_success, error=None if receipt.is_success else "Transaction failed")

        except Exception as e:
            return ContractResult(success=False, error=str(e))

    async def get_agreement(self, agreement_id: str) -> Optional[Agreement]:
        """
        Get agreement details

        Args:
            agreement_id: Agreement ID

        Returns:
            Agreement object or None if not found
        """
        if not self.contract_address:
            return None

        try:
            # Query contract storage
            # In production, use proper contract query mechanism
            # This is a simplified version
            return None  # Would implement with proper metadata
        except Exception as e:
            print(f"Error getting agreement: {e}")
            return None

    async def get_milestone(
        self,
        agreement_id: str,
        milestone_index: int
    ) -> Optional[Milestone]:
        """
        Get milestone details

        Args:
            agreement_id: Agreement ID
            milestone_index: Index of the milestone

        Returns:
            Milestone object or None if not found
        """
        if not self.contract_address:
            return None

        try:
            # Query contract storage
            return None  # Would implement with proper metadata
        except Exception as e:
            print(f"Error getting milestone: {e}")
            return None

    async def get_milestone_count(self, agreement_id: str) -> int:
        """
        Get milestone count for an agreement

        Args:
            agreement_id: Agreement ID

        Returns:
            Number of milestones
        """
        if not self.contract_address:
            return 0

        try:
            # Query contract storage
            return 0  # Would implement with proper metadata
        except Exception as e:
            print(f"Error getting milestone count: {e}")
            return 0

    def _encode_call(self, method: str, *args) -> bytes:
        """
        Encode contract call data

        In production, this would use the contract metadata to properly encode the call.
        For now, returns a placeholder.
        """
        # TODO: Implement proper encoding using contract metadata
        return b''

    def _extract_agreement_id(self, receipt) -> Optional[str]:
        """Extract agreement ID from transaction receipt events"""
        # TODO: Implement event parsing
        return None

File: packages/rust/src/contracts/escrow.rs
//! GenericEscrow contract client

use super::types::*;
use anyhow::Result;
use sp_core::sr25519::Pair;
use subxt::{OnlineClient, PolkadotConfig};

/// Client for interacting with GenericEscrow smart contract
///
/// Provides methods to interact with the GenericEscrow contract
/// for milestone-based payment escrow with dispute resolution.
///
/// # Example
///
/// ```no_run
/// use glin_sdk::contracts::{EscrowContract, CreateAgreementParams};
/// use sp_core::crypto::AccountId32;
///
/// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
/// let client = subxt::OnlineClient::new().await?;
/// let escrow = EscrowContract::new(client, "5Escrow...".parse()?);
///
/// let params = CreateAgreementParams {
///     provider: "5Provider...".parse()?,
///     milestone_descriptions: vec!["Design".to_string(), "Development".to_string()],
///     milestone_amounts: vec![500_000_000_000_000_000_000, 1_500_000_000_000_000_000_000],
///     milestone_deadlines: vec![1234567890, 1234567890],
///     dispute_timeout: 1234567890,
///     oracle: None,
///     value: 2_000_000_000_000_000_000_000,
/// };
///
/// // let result = escrow.create_agreement(params, &keypair).await?;
/// # Ok(())
/// # }
/// ```
pub struct EscrowContract {
    client: OnlineClient<PolkadotConfig>,
    contract_address: AccountId,
}

impl EscrowContract {
    /// Create a new escrow contract client
    pub fn new(client: OnlineClient<PolkadotConfig>, contract_address: AccountId) -> Self {
        Self {
            client,
            contract_address,
        }
    }

    /// Update contract address
    pub fn set_contract_address(&mut self, address: AccountId) {
        self.contract_address = address;
    }

    /// Create a new escrow agreement
    ///
    /// # Arguments
    ///
    /// * `params` - Agreement parameters
    /// * `signer` - Keypair for signing the transaction
    ///
    /// # Returns
    ///
    /// Returns the agreement ID on success
    pub async fn create_agreement(
        &self,
        params: CreateAgreementParams,
        signer: &Pair,
    ) -> Result<ContractResult<u128>> {
        // In production, this would:
        // 1. Encode the contract call using metadata
        // 2. Create a Contracts::call extrinsic
        // 3. Sign and submit the transaction
        // 4. Parse events to extract agreement_id

        // Placeholder implementation
        Ok(ContractResult::err("Not implemented - requires contract metadata"))
    }

    /// Mark a milestone as completed (by provider)
    ///
    /// # Arguments
    ///
    /// * `agreement_id` - Agreement ID
    /// * `milestone_index` - Index of the milestone
    /// * `signer` - Keypair for signing the transaction
    pub async fn complete_milestone(
        &self,
        agreement_id: u128,
        milestone_index: u32,
        signer: &Pair,
    ) -> Result<ContractResult<()>> {
        Ok(ContractResult::err("Not implemented - requires contract metadata"))
    }

    /// Approve milestone and release funds (by client or oracle)
    ///
    /// # Arguments
    ///
    /// * `agreement_id` - Agreement ID
    /// * `milestone_index` - Index of the milestone
    /// * `signer` - Keypair for signing the transaction
    pub async fn approve_and_release(
        &self,
        agreement_id: u128,
        milestone_index: u32,
        signer: &Pair,
    ) -> Result<ContractResult<()>> {
        Ok(ContractResult::err("Not implemented - requires contract metadata"))
    }

    /// Raise a dispute for a milestone
    ///
    /// # Arguments
    ///
    /// * `agreement_id` - Agreement ID
    /// * `milestone_index` - Index of the milestone
    /// * `signer` - Keypair for signing the transaction
    pub async fn raise_dispute(
        &self,
        agreement_id: u128,
        milestone_index: u32,
        signer: &Pair,
    ) -> Result<ContractResult<()>> {
        Ok(ContractResult::err("Not implemented - requires contract metadata"))
    }

    /// Resolve a dispute (by oracle or after timeout)
    ///
    /// # Arguments
    ///
    /// * `agreement_id` - Agreement ID
    /// * `milestone_index` - Index of the milestone
    /// * `release_to_provider` - Whether to release funds to provider
    /// * `signer` - Keypair for signing the transaction
    pub async fn resolve_dispute(
        &self,
        agreement_id: u128,
        milestone_index: u32,
        release_to_provider: bool,
        signer: &Pair,
    ) -> Result<ContractResult<()>> {
        Ok(ContractResult::err("Not implemented - requires contract metadata"))
    }

    /// Get agreement details
    ///
    /// # Arguments
    ///
    /// * `agreement_id` - Agreement ID
    ///
    /// # Returns
    ///
    /// Returns the agreement or None if not found
    pub async fn get_agreement(&self, agreement_id: u128) -> Result<Option<Agreement>> {
        // In production, this would query contract storage
        Ok(None)
    }

    /// Get milestone details
    ///
    /// # Arguments
    ///
    /// * `agreement_id` - Agreement ID
    /// * `milestone_index` - Index of the milestone
    ///
    /// # Returns
    ///
    /// Returns the milestone or None if not found
    pub async fn get_milestone(
        &self,
        agreement_id: u128,
        milestone_index: u32,
    ) -> Result<Option<Milestone>> {
        // In production, this would query contract storage
        Ok(None)
    }

    /// Get milestone count for an agreement
    ///
    /// # Arguments
    ///
    /// * `agreement_id` - Agreement ID
    ///
    /// # Returns
    ///
    /// Returns the number of milestones
    pub async fn get_milestone_count(&self, agreement_id: u128) -> Result<u32> {
        // In production, this would query contract storage
        Ok(0)
    }

    /// Get all milestones for an agreement (convenience method)
    ///
    /// # Arguments
    ///
    /// * `agreement_id` - Agreement ID
    ///
    /// # Returns
    ///
    /// Returns a vector of all milestones
    pub async fn get_all_milestones(&self, agreement_id: u128) -> Result<Vec<Milestone>> {
        let count = self.get_milestone_count(agreement_id).await?;
        let mut milestones = Vec::with_capacity(count as usize);

        for i in 0..count {
            if let Some(milestone) = self.get_milestone(agreement_id, i).await? {
                milestones.push(milestone);
            }
        }

        Ok(milestones)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_contract_result() {
        let ok_result: ContractResult<u128> = ContractResult::ok(42);
        assert!(ok_result.success);
        assert_eq!(ok_result.data, Some(42));
        assert!(ok_result.error.is_none());

        let err_result: ContractResult<u128> = ContractResult::err("Test error");
        assert!(!err_result.success);
        assert!(err_result.data.is_none());
        assert_eq!(err_result.error, Some("Test error".to_string()));
    }
}
